core/jobs/src/main/java/com/cognifide/aet/job/common/utils/Sampler.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * AET
+ *
+ * Copyright (C) 2013 Cognifide Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.cognifide.aet.job.common.utils;
+
+import java.util.LinkedList;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Sampler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Sampler.class);
+
+  /**
+   * Collects values from supplier in specified periods of time, and compares last n samples every
+   * iteration. If all n samples are equal(), returns the value. If last n samples don't match
+   * before max iterations threshold is reached, returns last collected sample.
+   *
+   * @param samplesSupplier supplier of value to wait for,
+   * @param samplingPeriod milliseconds period between taking each sample,
+   * @param sampleQueueSize defines the last n elements that are to be compared,
+   * @param maxSamplesThreshold max number of samples before return
+   * @return true if pattern is empty or if its match with given value, false otherwise
+   */
+  public static <T> T waitForValue(Supplier<T> samplesSupplier, int samplingPeriod,
+      int sampleQueueSize, int maxSamplesThreshold) {
+    LinkedList<T> samplesQueue = new LinkedList<>();
```

Comment:
Did you consider using `org.apache.commons.collections.buffer.CircularFifoBuffer` ? I think you can use it here to simplify your code :)